src/Kestrel.Core/Internal/Http/HttpProtocol.cs

```diff
@@ -39,12 +39,17 @@ public abstract partial class HttpProtocol : IHttpResponseControl
         private Stack<KeyValuePair<Func<object, Task>, object>> _onStarting;
         private Stack<KeyValuePair<Func<object, Task>, object>> _onCompleted;
 
-        private volatile int _ioCompleted;
+        private object _abortLock = new object();
+        private volatile bool _requestAborted;
+        private bool _preventRequestAbortedCancellation;
```

Comment:
If we were to do something like that, I'd add a RequestProcessingStatus.ResponseCompleted status. We would still need this flag though, because this isn't set when the response is completed if the request was already aborted.